Test multiple resolvers via matrix

stackptr opened the pull request:
Runs the `test` job against multiple Stack resolvers using a matrix strategy.

Motivated by #6, to exercise each case of the preprocessor directive that tests the version of `persistent`. LTS 18.0 pulls in the earlier version while the current 18.25 tests the later version.

Also, pins the version of `stack-action` and bumps it and `stack-cache-action` to latest version.

stackptr committed:
Test multiple resolvers via matrix

Runs the `test` job against multiple Stack resolvers using a matrix strategy.

Motivated by #6, to exercise each case of the preprocessor directive that tests the version of `persistent`. LTS 18.0 pulls in the earlier version while the current 18.25 tests the later version.

Also, pins the version of `stack-action` and bumps it and `stack-cache-action` to latest version.

stackptr commented:
Should `aws-xray-client-persistent` be bumped to version `0.1.0.3` or `0.1.1`? I'm not familiar with this versioning scheme.

pbrisbin commented on .github/workflows/ci.yml:
```diff
@@ -8,9 +8,21 @@ on:
+      matrix:
+        stack-yaml:
+          - stack.yaml
+          - stack-lts-18.0.yaml
+          - stack-lts-18.25.yaml
```
While these exercise the persistent thing, they're both the same GHC version. We like to use the matrix to hit various GHCs -- which should also exercise the persistent thing.

```
GHC 8.10  => lts-18.26
GHC 8.8   => lts-16.31
GHC 8.6   => lts-14.27
GHC 8.4   => lts-12.26
GHC 8.2   => lts-11.22
GHC 8.0   => lts-9.21
GHC 7.10  => lts-6.35
GHC 7.8   => lts-2.22
```

I typically do 12, 14, and 16. But if the older ones aren't easy to get compiling, just 16 in addition to the default is OK with me. (I also don't typically duplicate the current `stack.yaml`, so no 18 necessary.)

pbrisbin commented on stack-lts-18.0.yaml:
```diff
@@ -0,0 +1,30 @@
+- aws-xray-client-wai
+- aws-xray-client-persistent
+
+ghc-options:
```
I might not worry about this. Let our warning-preferences be `stack.yaml`-only to just cut down on the amount of copy/paste.

pbrisbin commented on .github/workflows/ci.yml:
```diff
@@ -8,9 +8,21 @@ on:
+          stack-yaml: ${{ matrix.stack-yaml }}
+      - uses: freckle/stack-action@v3.1.4
         with:
           weeder: false
```
You can remove this. The new version of the Action doesn't run weeder. You'll probably want to add an `hlint` Job too.

pbrisbin reviewed (changes requested):
Thanks for digging into this! Just blocking on picking different actual resolvers, to maximize their value.

pbrisbin commented:
For deciding the next version, you should follow [this guide](https://github.com/freckle/guides/blob/main/haskell-open-source.md#versioning).

stackptr committed:
Remove redundant `ghc-options`

stackptr committed:
Bump version

stackptr committed:
Remove duplicate LTS

stackptr committed:
Use LTSes with different versions of `ghc`

stackptr committed:
Wrap import in preprocessor directive

stackptr committed:
Remove test on LTS-12.26

It fails and seems low priority to support

stackptr committed:
Avoid redundant imports

stackptr committed:
Remove weeder, add lint job

stackptr commented:
I updated the branch protection rules to remove the required `test` status check (which no longer runs with this change) and added the three checks above as required.
